src/piadm.sh

```diff
@@ -0,0 +1,741 @@
+    # we only really care about boot.catalog and boot.  These may be remounted
+    # as read-only, so we can't do mv.
+
+    if [[ "$iso" == "yes" ]]; then
+	# Match-check boot stamp and platform stamp.
+	pstamp=$(cat ${tdir}/mnt/platform/etc/version/platform)
+	if [[ "$stamp" != "$pstamp" ]];	then
+	    umount ${tdir}/mnt
+	    /bin/rm -rf ${tdir}
+	    err "Boot stamp $stamp mismatches platform stamp $pstamp"
+	fi
+
+	if [[ -e /${bootfs}/boot-${stamp} ]]; then
+	    umount ${tdir}/mnt
+	    /bin/rm -rf ${tdir}
+	    echo "PI-stamp $stamp has boot bits already on /${bootfs}"
+	    err "Use   piadm remove $stamp   to remove any old copies."
+	fi
+	mkdir /${bootfs}/boot-${stamp}
+	tar -cf - -C ${tdir}/mnt/boot . | tar -xf - -C /${bootfs}/boot-${stamp}
+    fi
+
+    if [[ -e /${bootfs}/platform-${stamp} ]]; then
+	if [[ "$iso" == "yes" ]]; then
+	    umount ${tdir}/mnt
+	fi
+	/bin/rm -rf ${tdir}
+	echo "PI-stamp $stamp appears to be already on /${bootfs}"
+	err "Use   piadm remove $stamp   to remove any old copies."
+    fi
+    mkdir /${bootfs}/platform-${stamp}
+    tar -cf - -C ${tdir}/mnt/platform . | \
+	tar -xf - -C /${bootfs}/platform-${stamp}
+
+    if [[ "$iso" == "yes" ]]; then
+	umount ${tdir}/mnt
+    fi
+    /bin/rm -rf ${tdir}
+
+    # Global variable for enablepool() usage...
+    installstamp=$stamp
+    return 0
+}
+
+list() {
+    if [[ $1 == "-H" ]]; then
+	pool=$2
+    else
+	printf "%-18s %-30s %-12s %-5s %-5s \n" "PI STAMP" \
+	   "BOOTABLE FILESYSTEM" "BOOT BITS?" "NOW" "NEXT"
+	pool=$1
+    fi
+
+    poolpresent $pool
+
+    getbootable
+    for bootfs in ${allbootfs[@]}; do
+	bfspool=$(echo $bootfs | awk -F/ '{print $1}')
+	if [[ "$pool" != "" && "$bfspool" != "$pool" ]]; then
+	    # If we specify a pool for listing, skip ones not in the pool.
+	    continue
+	fi
+	if [[ ! -L /$bootfs/platform ]]; then
+	    corrupt \
+		"WARNING: Bootable filesystem $bootfs has non-symlink platform"
+	fi
+	cd /$bootfs
+	bootbitsstamp=$(cat etc/version/boot)
+	bootstamp=$(cat platform/etc/version/platform)
+	mapfile -t pis \
+		< <(cd /$bootfs ; cat platform-*/etc/version/platform)
+	for pi in ${pis[@]}; do
+	    if [[ $activestamp == $pi ]]; then
+		active="yes"
+	    else
+		active="no"
+	    fi
+	    if [[ $bootstamp == $pi ]]; then
+		booting="yes"
+	    else
+		booting="no"
+	    fi
+	    if [[ $bootbitsstamp == $pi ]]; then
+		bootbits="next"
+	    elif [[ -d boot-$pi ]]; then
+		bootbits="available"
+	    else
+		bootbits="none"
+	    fi
+	    printf "%-18s %-30s %-12s %-5s %-5s\n" $pi $bootfs $bootbits \
+		$active $booting
+	done
+    done
+}
+
+update_boot_sectors() {
+    pool=$1
+    bootfs=$2
+
+    # XXX WARNING -- illumos#12894 will allow slogs.  We will need to
+    # alter the generation of boot_devices accordingly.
+    # Generate the pool's boot devices now, in case we did something
+    # hyper-clever for the pool.  s1 may be created, but not yet PCFS...
+    mapfile -t boot_devices < <(zpool list -v "$pool" | \
```

Comment:
it's unfortunate that we kidn of have to use zpool list, but can it at least be zpool list -vHP zones | awk -F\\t '{print $2}'

Reply:
Sorry, I forgot -P is a guarantee of parsability.